### ISSUE TITLE
Add self.name method to abstract class

### ISSUE THREAD

**Issue by rogozin** [rkrage/pg_party#13]
I've caught an ArgumentError when I call `model_name` method for my partition table.

```
class TMainDatum < ApplicationRecord
  range_partition_by :date
end
```

For the parent class it's ok: 
```
TMainDatum.model_name
=> #<ActiveModel::Name:0x007ffd671791b8
 @collection="t_main_data",
 @element="t_main_datum",
 @human="T main datum",
 @i18n_key=:t_main_datum,
 @klass=TMainDatum (call 'TMainDatum.connection' to establish a connection),
 @name="TMainDatum",
 @param_key="t_main_datum",
 @plural="t_main_data",
 @route_key="t_main_data",
 @singular="t_main_datum",
 @singular_route_key="t_main_datum">
```

and for the child class it raises an error:
```
TMainDatum.in_partition('t_main_data_2017_34').model_name
ArgumentError: Class name cannot be blank. You need to supply a name argument when anonymous class given
from /ruby/gems/2.3.0/gems/activemodel-5.1.4/lib/active_model/naming.rb:149:in `initialize
```

Maybe you could simply to add `self.name` to ModelDecorator, couldn't you?

```
def child_class(table_name) do 
  Class.new(__getobj__) do
    ...
    def self.name
      superclass.name
    end
  end
end
```

The real usage of `model_name` dug into activerecord-import: https://github.com/zdennis/activerecord-import/blob/master/lib/activerecord-import/import.rb#L650

**Comment by rkrage** [comment]
Hi @rogozin, thanks for reporting this.

Just curious, is there a specific reason you're calling `model_name` here? The `in_partition` method currently returns an `ActiveRecord::Relation`. The anonymous class wasn't really meant to be leaked and interacted with (outside of chaining query methods).

But I think you're right, delegating `name` to the parent class seems like a reasonable thing to do.

**Comment by rogozin** [comment]
Hi @rkrage, thanks for your answer 👍 

Frankly I'm using [activerecord-import](https://github.com/zdennis/activerecord-import) with ON CONFILCT expression, so in fact I'm calling 

```
TMainDatum.import huge_data_array, on_duplicate_key_update: { target_and_column_options } 
``` 
And of course it won't work because there are PG partition restrictions for on_conflict usage .

Eventually I'm targeting specific partition like this:

```
TMainDatum.in_partition('t_main_data_2017_34').import huge_data_array, on_duplicate_key_update: { target_and_column_options }
```

and it works well, but the only exception as I've described above :)

**Comment by rkrage** [comment]
Ah, okay. Honestly, I didn't think a whole lot about use cases for `in_partition`. Maybe it makes sense to change the API a little bit then - have it return the raw child class so users can interact with it directly?

I have a pull request up that fixes the bug (https://github.com/rkrage/pg_party/pull/14), but I'll try to revise it a bit to reflect this API change.

Sound good?

**Comment by rkrage** [comment]
Side note, if it sounds like `in_partition` will be a widely used method, we might want to consider adding a bit of validation (check that the partition is actually a child partition of the parent), and add some class caching (to reduce the amount of child classes that are dynamically created).

Sounds like that work should be in a separate PR though.

**Comment by rkrage** [comment]
@rogozin the fix should be included in version `0.6.0`
